### PR TITLE
[flutter_tools] fix tests with no native assets running native asset build

### LINKED ISSUE #135461: Linux native assets integration is flaky on devicelab
See https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20framework_tests_slow/5675/overview

https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8768955848268631089/+/u/run_test.dart_for_framework_tests_shard_and_subshard_slow/test_stdout

Why is this running on a unit test?

```
  Exception: Failed to find clang++ on PATH.
  package:flutter_tools/src/base/common.dart 10:3            throwToolExit
  package:flutter_tools/src/linux/native_assets.dart 219:5   cCompilerConfigLinux
  ===== asynchronous gap ===========================
  package:flutter_tools/src/linux/native_assets.dart 108:22  buildNativeAssetsLinux
  ===== asynchronous gap ===========================
  package:flutter_tools/src/test/test_compiler.dart 191:35   TestCompiler._onCompilationRequest
  ```

### PR DESCRIPTION
**Pull request by christopherfujino**
Fixes https://github.com/flutter/flutter/issues/135461

### CODE REVIEW

File: packages/flutter_tools/lib/src/native_assets.dart [hunk 1]
```diff
@@ -214,45 +215,45 @@ BuildMode nativeAssetsBuildMode(build_info.BuildMode buildMode) {
...
-Future<bool> isDisabledAndNoNativeAssets(NativeAssetsBuildRunner buildRunner) async {
-  if (featureFlags.isNativeAssetsEnabled) {
+Future<bool> nativeBuildRequired(NativeAssetsBuildRunner buildRunner) async {
+  if (await _hasNoPackageConfig(buildRunner)) {
     return false;
   }
   final List<Package> packagesWithNativeAssets = await buildRunner.packagesWithNativeAssets();
   if (packagesWithNativeAssets.isEmpty) {
-    return true;
+    return false;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/native_assets.dart:232]: This is the fix

**Comment by christopherfujino** [comment]
WDYT @dcharkes ?

**Comment by dcharkes** [comment]
I don't think I can push to your branch, here's how to fix the two failing tests:

```patch
From b22459fa905f5c579efc938446e52feb40da5ae6 Mon Sep 17 00:00:00 2001
From: Daco Harkes <dacoharkes@google.com>
Date: Tue, 26 Sep 2023 09:53:45 +0200
Subject: [PATCH] Fix failing tests

---
 .../build_system/targets/native_assets_test.dart              | 4 ++++
 .../test/general.shard/resident_runner_test.dart              | 2 +-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart b/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart
index 09fe96148d..39f5110d87 100644
--- a/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart
@@ -14,6 +14,7 @@ import 'package:flutter_tools/src/build_system/targets/native_assets.dart';
 import 'package:flutter_tools/src/features.dart';
 import 'package:flutter_tools/src/native_assets.dart';
 import 'package:native_assets_cli/native_assets_cli.dart' as native_assets_cli;
+import 'package:package_config/package_config_types.dart';
 
 import '../../../src/common.dart';
 import '../../../src/context.dart';
@@ -104,6 +105,9 @@ void main() {
       await createPackageConfig(iosEnvironment);
 
       final NativeAssetsBuildRunner buildRunner = FakeNativeAssetsBuildRunner(
+        packagesWithNativeAssetsResult: <Package>[
+          Package('foo', iosEnvironment.buildDir.uri)
+        ],
         buildResult: FakeNativeAssetsBuilderResult(assets: <native_assets_cli.Asset>[
           native_assets_cli.Asset(
             id: 'package:foo/foo.dart',
diff --git a/packages/flutter_tools/test/general.shard/resident_runner_test.dart b/packages/flutter_tools/test/general.shard/resident_runner_test.dart
index e24262fff8..2af0b71530 100644
--- a/packages/flutter_tools/test/general.shard/resident_runner_test.dart
+++ b/packages/flutter_tools/test/general.shard/resident_runner_test.dart
@@ -2420,7 +2420,7 @@ flutter:
         expect(buildRunner.buildInvocations, 0);
         expect(buildRunner.dryRunInvocations, 1);
         expect(buildRunner.hasPackageConfigInvocations, 1);
-        expect(buildRunner.packagesWithNativeAssetsInvocations, 0);
+        expect(buildRunner.packagesWithNativeAssetsInvocations, 1);
       }),
       overrides: <Type, Generator>{
         ProcessManager: () => FakeProcessManager.any(),
-- 
2.42.0.515.g380fc7ccd1-goog
```

File: packages/flutter_tools/lib/src/native_assets.dart [hunk 2]
```diff
@@ -214,45 +214,45 @@ BuildMode nativeAssetsBuildMode(build_info.BuildMode buildMode) {
...
+    );
+  }
+  return true;
 }
 
 /// Ensures that either this project has no native assets, or that native assets
 /// are supported on that operating system.
 ///
 /// Exits the tool if the above condition is not satisfied.
+// TODO(fujino): this does not appear to do what it says it does.
```
Comment (dcharkes) [inline packages/flutter_tools/lib/src/native_assets.dart:248]: wdym?

This test checks that either (A) there are no native assets at all, or (B) it throws that the OS is not supported. This function is only called from call sites where the OS is not supported yet. (And (C) it skips the check if there is no package config.)

**Review by dcharkes — APPROVED** [review]
Thanks @christopherfujino !

LGTM!

**Comment by christopherfujino** [comment]
> I don't think I can push to your branch, here's how to fix the two failing tests:
> 
> ```diff
> From b22459fa905f5c579efc938446e52feb40da5ae6 Mon Sep 17 00:00:00 2001
> From: Daco Harkes <dacoharkes@google.com>
> Date: Tue, 26 Sep 2023 09:53:45 +0200
> Subject: [PATCH] Fix failing tests
> 
> ---
>  .../build_system/targets/native_assets_test.dart              | 4 ++++
>  .../test/general.shard/resident_runner_test.dart              | 2 +-
>  2 files changed, 5 insertions(+), 1 deletion(-)
> 
> diff --git a/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart b/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart
> index 09fe96148d..39f5110d87 100644
> --- a/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart
> +++ b/packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart
> @@ -14,6 +14,7 @@ import 'package:flutter_tools/src/build_system/targets/native_assets.dart';
>  import 'package:flutter_tools/src/features.dart';
>  import 'package:flutter_tools/src/native_assets.dart';
>  import 'package:native_assets_cli/native_assets_cli.dart' as native_assets_cli;
> +import 'package:package_config/package_config_types.dart';
>  
>  import '../../../src/common.dart';
>  import '../../../src/context.dart';
> @@ -104,6 +105,9 @@ void main() {
>        await createPackageConfig(iosEnvironment);
>  
>        final NativeAssetsBuildRunner buildRunner = FakeNativeAssetsBuildRunner(
> +        packagesWithNativeAssetsResult: <Package>[
> +          Package('foo', iosEnvironment.buildDir.uri)
> +        ],
>          buildResult: FakeNativeAssetsBuilderResult(assets: <native_assets_cli.Asset>[
>            native_assets_cli.Asset(
>              id: 'package:foo/foo.dart',
> diff --git a/packages/flutter_tools/test/general.shard/resident_runner_test.dart b/packages/flutter_tools/test/general.shard/resident_runner_test.dart
> index e24262fff8..2af0b71530 100644
> --- a/packages/flutter_tools/test/general.shard/resident_runner_test.dart
> +++ b/packages/flutter_tools/test/general.shard/resident_runner_test.dart
> @@ -2420,7 +2420,7 @@ flutter:
>          expect(buildRunner.buildInvocations, 0);
>          expect(buildRunner.dryRunInvocations, 1);
>          expect(buildRunner.hasPackageConfigInvocations, 1);
> -        expect(buildRunner.packagesWithNativeAssetsInvocations, 0);
> +        expect(buildRunner.packagesWithNativeAssetsInvocations, 1);
>        }),
>        overrides: <Type, Generator>{
>          ProcessManager: () => FakeProcessManager.any(),
> -- 
> 2.42.0.515.g380fc7ccd1-goog
> ```

Ahh thanks. Btw, you SHOULD be able to push to my PR branches, as a maintainer.

**Pull request merged**